pkg/nsx/services/vpc/vpc.go

```diff
@@ -735,6 +745,84 @@ func (s *VPCService) GetConnectionTypeFromConnectionPath(connectionPath string)
+func (s *VPCService) GetNetworkStackFromProfile(profile *model.VpcConnectivityProfile) (v1alpha1.NetworkStackType, error) {
+	networkStack := v1alpha1.FullStackVPC
+	if profile == nil || profile.ParentPath == nil {
+		err := fmt.Errorf("profile or its parent path is nil")
+		log.Error(err, "Failed to parse NSX project in VpcConnectivityProfile")
+		return networkStack, err
+	}
+	org, project, err := common.NSXProjectPathToId(*profile.ParentPath)
+	if err != nil {
+		log.Error(err, "Failed to parse NSX project in VpcConnectivityProfile", "ProjectPath", *profile.ParentPath)
+		return networkStack, err
+	}
+	transitGatewayPath := *profile.TransitGatewayPath
+	parts := strings.Split(transitGatewayPath, "/")
+	transitGatewayId := parts[len(parts)-1]
+	state, err := s.NSXClient.TransitGatewayStateClient.Get(org, project, transitGatewayId, nil)
+	err = nsxutil.TransNSXApiError(err)
+	if err != nil {
+		return networkStack, err
+	}
+
+	if state.NetworkStack != nil {
+		switch *state.NetworkStack {
+		case model.TransitGatewayState_NETWORK_STACK_FULL_STACK_VPC:
+			networkStack = v1alpha1.FullStackVPC
+		case model.TransitGatewayState_NETWORK_STACK_VLAN_BACKED_VPC:
+			networkStack = v1alpha1.VLANBackedVPC
+		}
+	}
+	return networkStack, nil
+}
+
+func (s *VPCService) GetVpcConnectivityProfilePathByVpcPath(vpcPath string) (string, error) {
+	// TODO, if needs to add a cache for it
+	VPCResourceInfo, err := common.ParseVPCResourcePath(vpcPath)
+	if err != nil {
+		log.Error(err, "Failed to parse VPC path", "VPC Path", vpcPath)
+		return "", err
+	}
+	// pre created VPC may have more than one attachment, list all the attachment and select the first one
+	vpcAttachmentsListResult, err := s.Service.NSXClient.VpcAttachmentClient.List(VPCResourceInfo.OrgID, VPCResourceInfo.ProjectID, VPCResourceInfo.VPCID, nil, nil, nil, nil, nil, nil)
+	if err != nil {
+		log.Error(err, "Failed to list VPC attachment", "VPC Path", vpcPath)
+		return "", err
+	}
+	vpcAttachments := vpcAttachmentsListResult.Results
+	if len(vpcAttachments) > 0 {
+		log.Debug("Found VPC attachment", "VPC Path", vpcPath, "VPC connectivity profile", vpcAttachments[0].VpcConnectivityProfile)
+		return *vpcAttachments[0].VpcConnectivityProfile, nil
+	} else {
+		log.Debug("No VPC attachment", "VPC Path", vpcPath)
+		return "", nil
+	}
+}
+
+// GetNetworkStackFromNC checks the transit gateway state to determine the network stack type.
+func (s *VPCService) GetNetworkStackFromNC(nc *v1alpha1.VPCNetworkConfiguration) (v1alpha1.NetworkStackType, error) {
+	profilePath := nc.Spec.VPCConnectivityProfile
+	if IsPreCreatedVPC(nc) {
+		profilePath, err := s.GetVpcConnectivityProfilePathByVpcPath(nc.Spec.VPC)
+		if err != nil {
+			log.Error(err, "Failed to get VPCConnectivityProfile path from pre-created VPC", "VPC Path", nc.Spec.VPC)
+			return v1alpha1.FullStackVPC, err
+		}
+		if profilePath == "" {
+			return v1alpha1.FullStackVPC, nil
```

Comment:
We will also support noNat/noLB case for tepless, why it return FullStackVPC if profile path is empty?

Reply:
update to checking the state API.
we may not always get profile path